clients/drcachesim/tools/record_view.h

```diff
@@ -0,0 +1,92 @@
+/* **********************************************************
+ * Copyright (c) 2024 Google, Inc.  All rights reserved.
+ * **********************************************************/
+
+/*
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions are met:
+ *
+ * * Redistributions of source code must retain the above copyright notice,
+ *   this list of conditions and the following disclaimer.
+ *
+ * * Redistributions in binary form must reproduce the above copyright notice,
+ *   this list of conditions and the following disclaimer in the documentation
+ *   and/or other materials provided with the distribution.
+ *
+ * * Neither the name of Google, Inc. nor the names of its contributors may be
+ *   used to endorse or promote products derived from this software without
+ *   specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ * ARE DISCLAIMED. IN NO EVENT SHALL VMWARE, INC. OR CONTRIBUTORS BE LIABLE
+ * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+ * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+ * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ * CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
+ * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH
+ * DAMAGE.
+ */
+
+#ifndef _RECORD_VIEW_H_
+#define _RECORD_VIEW_H_ 1
+
+#include <stdint.h>
+
+#include <string>
+
+#include "analysis_tool.h"
+#include "memtrace_stream.h"
+#include "trace_entry.h"
+
+namespace dynamorio {
+namespace drmemtrace {
+
+/**
```

Comment:
I have mixed feelings about adding this tool: part of me does not want to add it as we do not want to encourage users to create tools that operate on the trace_entry_t format.  We consider that format subject to change and to not be a public interface.  It has once-only data that requires caching that easily leads to errors if users operate on it.  We have already seen external converters such as a champsim converter prototype incorrectly use this format.

What I'm getting as is should this have Doxygen comments for public docs?

Reply:
Oh, I thought tools that operate on trace_entry_t were public for some reason.
My vote is for giving "record_view" the same visibility as "record_filter" (which is also a tool that operates on trace_entry_t).
"record_filter" has doxygen comments for public docs, so I followed the same for "record_view".

If tools that operate on trace_entry_t are not supposed to be public, then we should remove the public documentation for all of them and transform doxygen comments in regular comments.
(Separate PR for "record_filter"?)

Reply:
No, trace_entry_t is not meant to be a public interface with guaranteed stability.  The public interface is the memref_t record and reader library and the libraries on top of those like the analysis tools.  trace_entry_t is supposed to be treated more like a black box by users.  Its individual fields are not doxygen-commented.

Hence, providing a pretty printer seems to send the wrong message, encouraging users to peer inside the black box.  We would prefer things like converters to other simulator formats to use memref_t + reader_t whose abstraction layer hides a number of messy details.

The record_filter tool can be used while keeping trace_entry_t a black box, as basically all of its filters operate above the memref_t abstraction level: remove marker types; trim to timestamp values; apply cache filters.  So it is different from a trace_entry_t pretty printer.

Reply:
What are the use cases of the record_view tool? The one I had in mind was debugging invariant errors in the raw trace. But that is likely more a trace-dev use case, rather than for trace users.

Reply:
> The one I had in mind was debugging invariant errors in the raw trace.

You mean the underlying trace_entry_t file on disk and not the offline "raw" un-pre-processed files, right?

Reply:
Along that line of thought: I would see a trace_entry_t pretty-printer along the same lines as a raw pretty-printer: used only by trace-internal devs and not by any trace gatherers/users.

Reply:
> You mean the underlying trace_entry_t file on disk and not the offline "raw" un-pre-processed files, right?

Well both can be useful for debugging invariant errors. I agree that both are internal use cases. But does that mean we should've add the record_view at all? Or maybe just make it a separate executable that's visibly marked internal (instead of supporting `-simulator_type record_view`).

Reply:
Making it clear it's internal seems best; maybe even in the name: undocumented publicly, but can trigger via `-simulator_type internal_record_view`.  And it prints a header every time saying "The internal record format is not a public interface and is subject to change without notice; meant for debugging only; please do not rely on the details here" or somesuch.

Reply:
Adding "internal" to the name SGTM.

Reply:
Ok, I think the final decision here is to add the "warning header" (done) and renaming record_view to internal_record_view (done).